File: src/shared/api/documents.ts
import { invokeLogged, isTauriRuntime } from "./runtime";
import { createContentHash, createLocalId } from "../id";
import { cloneWorkspaceSnapshot } from "../mock/workspace";
import {
  browserMock,
  captureBrowserDocumentHistory,
  clearBrowserDocumentHistory,
  createMockImagePreviewDataUrl,
  ensureParentFolderExistsForMock,
  joinRelativePath,
  migrateBrowserDocumentHistoryTarget,
  normalizeFolderPath,
  replaceFileNameInPath,
  validateNewTextDocumentFileNameForMock,
  validateTextDocumentFileNameForMock,
} from "../mock/browser";
import {
  DocumentPreview,
  ExportFileResult,
  ExportFormat,
  ExportTargetKind,
  KnowledgeBase,
  Note,
  WorkspaceDocument,
  WorkspaceSnapshot,
} from "../types";

/** 在用户点击的目录下新建空白 TXT；桌面端会立即创建真实文件。 */
export async function createDocument(
  snapshot: WorkspaceSnapshot,
  knowledgeBaseId: string,
  parentPath: string,
  fileName: string,
): Promise<WorkspaceSnapshot> {
  if (!isTauriRuntime()) {
    const nextSnapshot = cloneWorkspaceSnapshot(snapshot);
    const knowledgeBase = nextSnapshot.knowledgeBases.find((item) => item.id === knowledgeBaseId);

    if (!knowledgeBase) {
      return nextSnapshot;
    }

    const safeFileName = validateNewTextDocumentFileNameForMock(fileName);
    const normalizedParentPath = normalizeFolderPath(parentPath);
    ensureParentFolderExistsForMock(nextSnapshot, knowledgeBaseId, normalizedParentPath);
    const existingPaths = new Set([
      ...nextSnapshot.notes.filter((note) => note.knowledgeBaseId === knowledgeBaseId).map((note) => note.path),
      ...nextSnapshot.documents.filter((document) => document.knowledgeBaseId === knowledgeBaseId).map((document) => document.path),
    ]);
    const nextPath = joinRelativePath(normalizedParentPath, safeFileName);

    // 浏览器 fallback 也模拟桌面文件系统的同目录不可覆盖规则。
    if (existingPaths.has(nextPath)) {
      throw new Error("目标文件已存在，已阻止覆盖。");
    }

    const newDocument: WorkspaceDocument = {
      id: createLocalId("document"),
      knowledgeBaseId,
      title: safeFileName.replace(/\.txt$/i, ""),
      path: nextPath,
      fileType: "txt",
      content: "",
      contentHash: createContentHash(""),
      updatedAt: "刚刚",
      previewAvailable: false,
    };

    nextSnapshot.documents = [newDocument, ...nextSnapshot.documents];
    browserMock.documentDiskContents.set(newDocument.id, newDocument.content ?? "");
    nextSnapshot.knowledgeBases = nextSnapshot.knowledgeBases.map((item) =>
      item.id === knowledgeBaseId
        ? {
            ...item,
            documentCount: item.documentCount + 1,
            updatedAt: "刚刚",
            scanReport: item.scanReport
              ? {
                  ...item.scanReport,
                  scannedFileCount: item.scanReport.scannedFileCount + 1,
                  scannedByType: {
                    ...item.scanReport.scannedByType,
                    txt: (item.scanReport.scannedByType.txt ?? 0) + 1,
                  },
                }
              : item.scanReport,
          }
        : item,
    );
    nextSnapshot.activeKnowledgeBaseId = knowledgeBaseId;
    nextSnapshot.activeNoteId = "";
    nextSnapshot.activeDocumentId = newDocument.id;

    return nextSnapshot;
  }

  return invokeLogged<WorkspaceSnapshot>("create_document", {
    payload: { snapshot, knowledgeBaseId, parentPath, fileName },
  });
}

/** 保存当前 TXT 文档正文；桌面端会执行 hash 冲突检测和原子写入。 */
export async function saveDocumentContent(
  snapshot: WorkspaceSnapshot,
  documentId: string,
  content: string,
  expectedHash: string,
): Promise<WorkspaceSnapshot> {
  if (!isTauriRuntime()) {
    const nextSnapshot = cloneWorkspaceSnapshot(snapshot);
    const document = nextSnapshot.documents.find((item) => item.id === documentId);

    if (!document) {
      throw new Error("找不到要保存的文档。");
    }

    if (document.fileType !== "txt") {
      throw new Error("只有 TXT 文档支持保存。");
    }

    const diskContent = browserMock.documentDiskContents.get(document.id) ?? document.content ?? "";
    const diskHash = createContentHash(diskContent);

    if (diskHash !== expectedHash) {
      throw new Error("目标文件已被外部修改，已阻止保存。请重新扫描后再编辑。");
    }

    captureBrowserDocumentHistory({
      targetKind: "document",
      knowledgeBaseId: document.knowledgeBaseId,
      targetId: document.id,
      relativePath: document.path,
      title: document.title,
      fileType: "txt",
      content: diskContent,
      source: "manual-save",
    });

    browserMock.documentDiskContents.set(document.id, content);
    nextSnapshot.documents = nextSnapshot.documents.map((item) =>
      item.id === documentId ? { ...item, content, contentHash: createContentHash(content), updatedAt: "刚刚" } : item,
    );
    nextSnapshot.activeNoteId = "";
    nextSnapshot.activeDocumentId = documentId;

    return nextSnapshot;
  }

  return invokeLogged<WorkspaceSnapshot>("save_document_content", { payload: { snapshot, documentId, content, expectedHash } });
}

/** 重命名当前 TXT 文档；桌面端调用真实 Tauri 文件系统能力。 */
export async function renameDocument(
  snapshot: WorkspaceSnapshot,
  documentId: string,
  nextFileName: string,
): Promise<WorkspaceSnapshot> {
  if (!isTauriRuntime()) {
    const nextSnapshot = cloneWorkspaceSnapshot(snapshot);
    const document = nextSnapshot.documents.find((item) => item.id === documentId);

    if (!document) {
      throw new Error("找不到要重命名的文档。");
    }

    if (document.fileType !== "txt") {
      throw new Error("只有 TXT 文档支持重命名。");
    }

    const safeFileName = validateTextDocumentFileNameForMock(nextFileName);
    const nextPath = replaceFileNameInPath(document.path, safeFileName);
    const isPathTaken =
      nextSnapshot.notes.some((item) => item.knowledgeBaseId === document.knowledgeBaseId && item.path === nextPath) ||
      nextSnapshot.documents.some(
        (item) => item.knowledgeBaseId === document.knowledgeBaseId && item.id !== document.id && item.path === nextPath,
      );

    // 文件系统同目录不能出现同名文件，不区分它属于 note 还是 document 模型。
    if (isPathTaken) {
      throw new Error("目标文件名已存在，已阻止覆盖。");
    }

    const nextDocumentId = createLocalId("document-renamed");
    const diskContent = browserMock.documentDiskContents.get(document.id) ?? document.content ?? "";
    const nextTitle = safeFileName.replace(/\.txt$/i, "");

    nextSnapshot.documents = nextSnapshot.documents.map((item) =>
      item.id === document.id
        ? {
            ...item,
            id: nextDocumentId,
            path: nextPath,
            title: nextTitle,
            updatedAt: "刚刚",
          }
        : item,
    );
    browserMock.documentDiskContents.delete(document.id);
    browserMock.documentDiskContents.set(nextDocumentId, diskContent);
    migrateBrowserDocumentHistoryTarget("document", document.id, nextDocumentId, nextPath, nextTitle);

    if (nextSnapshot.activeDocumentId === document.id) {
      nextSnapshot.activeDocumentId = nextDocumentId;
      nextSnapshot.activeNoteId = "";
    }

    return nextSnapshot;
  }

  return invokeLogged<WorkspaceSnapshot>("rename_document", { payload: { snapshot, documentId, nextFileName } });
}

/** 删除当前 TXT 文档到系统回收站；浏览器 fallback 只移除内存快照中的模拟文档。 */
export async function deleteDocument(
  snapshot: WorkspaceSnapshot,
  documentId: string,
  expectedHash: string,
): Promise<WorkspaceSnapshot> {
  if (!isTauriRuntime()) {
    const nextSnapshot = cloneWorkspaceSnapshot(snapshot);
    const document = nextSnapshot.documents.find((item) => item.id === documentId);

    if (!document) {
      throw new Error("找不到要删除的文档。");
    }

    if (document.fileType !== "txt") {
      throw new Error("只有 TXT 文档支持删除。");
    }

    // 与桌面 Tauri command 保持一致：删除前必须确认操作基于同一份文件版本。
    const diskContent = browserMock.documentDiskContents.get(document.id) ?? document.content ?? "";

    if (createContentHash(diskContent) !== expectedHash) {
      throw new Error("目标文件已被外部修改，已阻止删除。请重新扫描后再操作。");
    }

    nextSnapshot.documents = nextSnapshot.documents.filter((item) => item.id !== documentId);
    browserMock.documentDiskContents.delete(documentId);
    clearBrowserDocumentHistory("document", documentId);
    nextSnapshot.knowledgeBases = nextSnapshot.knowledgeBases.map((knowledgeBase) =>
      knowledgeBase.id === document.knowledgeBaseId
        ? {
            ...knowledgeBase,
            documentCount: Math.max(0, knowledgeBase.documentCount - 1),
            updatedAt: "刚刚",
            scanReport: knowledgeBase.scanReport
              ? {
                  ...knowledgeBase.scanReport,
                  scannedFileCount: Math.max(0, knowledgeBase.scanReport.scannedFileCount - 1),
                  scannedByType: {
                    ...knowledgeBase.scanReport.scannedByType,
                    txt: Math.max(0, (knowledgeBase.scanReport.scannedByType.txt ?? 0) - 1),
                  },
                }
              : knowledgeBase.scanReport,
          }
        : knowledgeBase,
    );

    if (nextSnapshot.activeDocumentId === documentId) {
      const sameKnowledgeBaseFallback = nextSnapshot.documents.find((item) => item.knowledgeBaseId === document.knowledgeBaseId);

      nextSnapshot.activeDocumentId = sameKnowledgeBaseFallback?.id ?? "";
      nextSnapshot.activeNoteId = sameKnowledgeBaseFallback ? "" : nextSnapshot.notes.find((item) => item.knowledgeBaseId === document.knowledgeBaseId)?.id ?? "";
    }

    return nextSnapshot;
  }

  return invokeLogged<WorkspaceSnapshot>("delete_document", { payload: { snapshot, documentId, expectedHash } });
}

/** 加载 DOCX/PDF/图片只读预览；TXT 直接使用快照中的 content。 */
export async function loadDocumentPreview(snapshot: WorkspaceSnapshot, documentId: string): Promise<DocumentPreview> {
  const document = snapshot.documents.find((item) => item.id === documentId);

  if (!document) {
    throw new Error("找不到要预览的文档。");
  }

  if (!isTauriRuntime()) {
    return {
      documentId: document.id,
      fileType: document.fileType,
      title: document.title,
      path: document.path,
      updatedAt: document.updatedAt,
      contentHash: document.contentHash,
      assetPath:
        document.fileType === "pdf"
          ? document.path
          : document.fileType === "image"
            ? createMockImagePreviewDataUrl(document.title)
            : undefined,
      blocks:
        document.fileType === "docx"
          ? [
              { type: "heading", text: document.title },
              { type: "paragraph", text: "这是浏览器开发态模拟的 DOCX 只读预览正文。" },
            ]
          : undefined,
    };
  }

  return invokeLogged<DocumentPreview>("load_document_preview", { payload: { snapshot, documentId } });
}

/** 导出当前打开文件；真实文件写入只允许在 Tauri 桌面端通过系统保存对话框完成。 */
export async function exportCurrentFile(
  snapshot: WorkspaceSnapshot,
  targetKind: ExportTargetKind,
  targetId: string,
  format: ExportFormat,
): Promise<ExportFileResult | null> {
  if (!isTauriRuntime()) {
    throw new Error("浏览器开发态不能导出本地文件，请在 Tauri 桌面端使用导出。");
  }

  return invokeLogged<ExportFileResult | null>("export_current_file", {
    payload: { snapshot, targetKind, targetId, format },
  });
}

File: src-tauri/src/commands/documents.rs
use super::common::*;

#[tauri::command]
pub async fn create_document(
    app: AppHandle,
    payload: CreateDocumentPayload,
) -> Result<WorkspaceSnapshot, String> {
    let started_at = Instant::now();
    let mut snapshot = payload.snapshot;
    let knowledge_base_index = snapshot
        .knowledge_bases
        .iter()
        .position(|knowledge_base| knowledge_base.id == payload.knowledge_base_id)
        .ok_or_else(|| "找不到要新建文档的知识库".to_owned())?;
    let knowledge_base = snapshot.knowledge_bases[knowledge_base_index].clone();

    if knowledge_base.status == "error" {
        return Err("当前知识库目录不可访问，无法新建 TXT 文档。".to_owned());
    }

    let root_path = PathBuf::from(&knowledge_base.path);
    let parent_path = payload.parent_path.unwrap_or_default();
    let file_name = payload.file_name;
    let relative_path = run_blocking("创建空白 TXT 文件", move || {
        storage::create_blank_text_document_file(&root_path, &parent_path, file_name.as_deref())
    })
    .await?;
    let created_relative_path = relative_path.clone();
    let document_id = storage::create_stable_document_id(&knowledge_base.id, &relative_path);
    let created_document_path = PathBuf::from(&knowledge_base.path).join(&relative_path);
    let updated_at = read_file_updated_at_or_now(
        &app,
        "create_document",
        &knowledge_base.id,
        "document",
        &document_id,
        &relative_path,
        &created_document_path,
    );
    let new_document = crate::domain::WorkspaceDocument {
        id: document_id.clone(),
        knowledge_base_id: knowledge_base.id.clone(),
        title: document_title_from_path(&relative_path),
        path: relative_path,
        file_type: "txt".to_owned(),
        updated_at,
        content_hash: storage::hash_content(""),
        content: Some(String::new()),
        preview_available: false,
    };

    snapshot.documents.insert(0, new_document);
    snapshot.knowledge_bases[knowledge_base_index].document_count += 1;
    snapshot.knowledge_bases[knowledge_base_index].updated_at = "刚刚".to_owned();
    if let Some(scan_report) = &mut snapshot.knowledge_bases[knowledge_base_index].scan_report {
        scan_report.scanned_file_count += 1;
        *scan_report
            .scanned_by_type
            .entry("txt".to_owned())
            .or_insert(0) += 1;
    }
    snapshot.active_knowledge_base_id = knowledge_base.id.clone();
    snapshot.active_note_id.clear();
    snapshot.active_document_id = document_id;
    normalize_active_entities(&mut snapshot, Some(&knowledge_base.id));

    logging::write_app_event_best_effort(
        &app,
        AppEventBuilder::new(
            AppLogLevel::Info,
            AppLogCategory::Editor,
            "create_document",
            "completed",
            "已创建 TXT 文档。",
        )
        .duration(started_at.elapsed())
        .knowledge_base_id(knowledge_base.id)
        .entity("document", snapshot.active_document_id.clone())
        .relative_path(created_relative_path),
    );

    Ok(snapshot)
}

/** 用户在目录树的指定目录下新建文件夹，成功后只更新目录快照不切换当前笔记。 */
#[tauri::command]
pub async fn create_folder(
    app: AppHandle,
    payload: CreateFolderPayload,
) -> Result<WorkspaceSnapshot, String> {
    let started_at = Instant::now();
    let mut snapshot = payload.snapshot;
    let knowledge_base_index = snapshot
        .knowledge_bases
        .iter()
        .position(|knowledge_base| knowledge_base.id == payload.knowledge_base_id)
        .ok_or_else(|| "找不到要新建文件夹的知识库".to_owned())?;
    let knowledge_base = snapshot.knowledge_bases[knowledge_base_index].clone();

    if knowledge_base.status == "error" {
        return Err("当前知识库目录不可访问，无法新建文件夹。".to_owned());
    }

    let root_path = PathBuf::from(&knowledge_base.path);
    let parent_path = payload.parent_path;
    let folder_name = payload.folder_name;
    let relative_path = run_blocking("创建文件夹", move || {
        storage::create_folder(&root_path, &parent_path, &folder_name)
    })
    .await?;
    let created_relative_path = relative_path.clone();
    let folder_id = storage::create_stable_folder_id(&knowledge_base.id, &relative_path);
    let folder_entry = FolderEntry {
        id: folder_id.clone(),
        knowledge_base_id: knowledge_base.id.clone(),
        name: folder_name_from_path(&relative_path),
        path: relative_path,
        updated_at: "刚刚".to_owned(),
    };

    // 快照可能来自旧版本或浏览器 fallback，追加前去重，避免同一目录显示两次。
    snapshot.folders.retain(|folder| {
        // 只在当前知识库内去重新建目录，不能影响其他知识库中同名相对目录。
        folder.knowledge_base_id != knowledge_base.id
            || (folder.id != folder_entry.id && folder.path != folder_entry.path)
    });
    snapshot.folders.push(folder_entry);
    snapshot.knowledge_bases[knowledge_base_index].updated_at = "刚刚".to_owned();
    snapshot.active_knowledge_base_id = knowledge_base.id.clone();
    normalize_active_entities(&mut snapshot, Some(&knowledge_base.id));

    logging::write_app_event_best_effort(
        &app,
        AppEventBuilder::new(
            AppLogLevel::Info,
            AppLogCategory::Editor,
            "create_folder",
            "completed",
            "已创建文件夹。",
        )
        .duration(started_at.elapsed())
        .knowledge_base_id(knowledge_base.id)
        .entity("folder", folder_id)
        .relative_path(created_relative_path),
    );

    Ok(snapshot)
}

/** 重命名 Markdown 文件，只修改文件名，并同步更新快照与会话引用。 */

#[tauri::command]
pub async fn rename_document(
    app: AppHandle,
    payload: RenameDocumentPayload,
) -> Result<WorkspaceSnapshot, String> {
    let started_at = Instant::now();
    let mut snapshot = payload.snapshot;
    let document_index = snapshot
        .documents
        .iter()
        .position(|document| document.id == payload.document_id)
        .ok_or_else(|| "找不到要重命名的文档".to_owned())?;
    let previous_document = snapshot.documents[document_index].clone();

    if previous_document.file_type != "txt" {
        return Err("只有 TXT 文档支持重命名。".to_owned());
    }

    let knowledge_base = snapshot
        .knowledge_bases
        .iter()
        .find(|item| item.id == previous_document.knowledge_base_id)
        .cloned()
        .ok_or_else(|| "找不到文档所属知识库".to_owned())?;
    let root_path = PathBuf::from(&knowledge_base.path);
    let current_relative_path = previous_document.path.clone();
    let next_file_name = payload.next_file_name;
    let (next_relative_path, current_content, current_hash) =
        run_blocking("重命名 TXT 文件", move || {
            storage::rename_text_document_file(&root_path, &current_relative_path, &next_file_name)
        })
        .await?;
    let next_document_id =
        storage::create_stable_document_id(&knowledge_base.id, &next_relative_path);
    let next_document_title = document_title_from_path(&next_relative_path);
    let history_migrate_app = app.clone();
    let history_previous_document_id = payload.document_id.clone();
    let history_next_document_id = next_document_id.clone();
    let history_knowledge_base_id = knowledge_base.id.clone();
    let history_next_relative_path = next_relative_path.clone();
    let history_next_title = next_document_title.clone();

    if let Err(_) = run_blocking("迁移 TXT 历史记录", move || {
        storage::migrate_document_history_target(
            &history_migrate_app,
            "document",
            &history_previous_document_id,
            &history_next_document_id,
            &history_knowledge_base_id,
            &history_next_relative_path,
            &history_next_title,
        )
    })
    .await
    {
        logging::write_app_event_best_effort(
            &app,
            AppEventBuilder::new(
                AppLogLevel::Warn,
                AppLogCategory::Editor,
                "document_history_migration",
                "failed",
                "TXT 重命名后历史记录迁移失败。",
            )
            .duration(started_at.elapsed())
            .knowledge_base_id(knowledge_base.id.clone())
            .entity("document", next_document_id.clone())
            .relative_path(next_relative_path.clone())
            .metadata(json!({
                "targetKind": "document",
                "failureKind": "migration_failed",
            })),
        );
    }
    let next_document_path = PathBuf::from(&knowledge_base.path).join(&next_relative_path);
    let updated_at = read_file_updated_at_or_now(
        &app,
        "rename_document",
        &knowledge_base.id,
        "document",
        &next_document_id,
        &next_relative_path,
        &next_document_path,
    );

    snapshot.documents[document_index].id = next_document_id.clone();
    snapshot.documents[document_index].title = next_document_title;
    snapshot.documents[document_index].path = next_relative_path;
    snapshot.documents[document_index].content = Some(current_content);
    snapshot.documents[document_index].content_hash = current_hash;
    snapshot.documents[document_index].updated_at = updated_at;

    if snapshot.active_document_id == payload.document_id {
        snapshot.active_document_id = next_document_id;
        snapshot.active_note_id.clear();
    }

    logging::write_app_event_best_effort(
        &app,
        AppEventBuilder::new(
            AppLogLevel::Info,
            AppLogCategory::Editor,
            "rename_document",
            "completed",
            "已重命名 TXT 文档。",
        )
        .duration(started_at.elapsed())
        .knowledge_base_id(knowledge_base.id)
        .entity("document", snapshot.documents[document_index].id.clone())
        .relative_path(snapshot.documents[document_index].path.clone()),
    );

    Ok(snapshot)
}

/** 删除 Markdown 文件到系统回收站，并从快照中移除笔记和相关会话引用。 */

#[tauri::command]
pub async fn delete_document(
    app: AppHandle,
    payload: DeleteDocumentPayload,
) -> Result<WorkspaceSnapshot, String> {
    let started_at = Instant::now();
    let mut snapshot = payload.snapshot;
    let document_index = snapshot
        .documents
        .iter()
        .position(|document| document.id == payload.document_id)
        .ok_or_else(|| "找不到要删除的文档".to_owned())?;
    let document = snapshot.documents[document_index].clone();

    if document.file_type != "txt" {
        return Err("只有 TXT 文档支持删除。".to_owned());
    }

    let knowledge_base_index = snapshot
        .knowledge_bases
        .iter()
        .position(|item| item.id == document.knowledge_base_id)
        .ok_or_else(|| "找不到文档所属知识库".to_owned())?;
    let knowledge_base = snapshot.knowledge_bases[knowledge_base_index].clone();
    let root_path = PathBuf::from(&knowledge_base.path);
    let relative_path = document.path.clone();
    let expected_hash = payload.expected_hash;

    run_blocking("删除 TXT 文件", move || {
        storage::trash_text_document_file(&root_path, &relative_path, &expected_hash)
    })
    .await?;

    clear_document_history_after_delete_best_effort(
        &app,
        "document",
        document.id.clone(),
        knowledge_base.id.clone(),
        document.path.clone(),
        started_at,
    )
    .await;

    snapshot.documents.remove(document_index);
    snapshot.knowledge_bases[knowledge_base_index].document_count = snapshot.knowledge_bases
        [knowledge_base_index]
        .document_count
        .saturating_sub(1);
    snapshot.knowledge_bases[knowledge_base_index].updated_at = "刚刚".to_owned();

    if let Some(scan_report) = &mut snapshot.knowledge_bases[knowledge_base_index].scan_report {
        scan_report.scanned_file_count = scan_report.scanned_file_count.saturating_sub(1);
        if let Some(txt_count) = scan_report.scanned_by_type.get_mut("txt") {
            *txt_count = txt_count.saturating_sub(1);
        }
    }

    if snapshot.active_document_id == payload.document_id {
        snapshot.active_document_id.clear();
    }

    normalize_active_entities(&mut snapshot, Some(&knowledge_base.id));

    logging::write_app_event_best_effort(
        &app,
        AppEventBuilder::new(
            AppLogLevel::Info,
            AppLogCategory::Editor,
            "delete_document",
            "completed",
            "已将 TXT 文档移入回收站。",
        )
        .duration(started_at.elapsed())
        .knowledge_base_id(knowledge_base.id)
        .entity("document", document.id)
        .relative_path(document.path),
    );

    Ok(snapshot)
}

/** 保存当前笔记正文，校验知识库边界和文件 hash 后原子写回 Markdown。 */

#[tauri::command]
pub async fn save_document_content(
    app: AppHandle,
    payload: SaveDocumentContentPayload,
) -> Result<WorkspaceSnapshot, String> {
    let started_at = Instant::now();
    let mut snapshot = payload.snapshot;
    let document_index = snapshot
        .documents
        .iter()
        .position(|document| document.id == payload.document_id)
        .ok_or_else(|| "找不到要保存的文档".to_owned())?;

    if snapshot.documents[document_index].file_type != "txt" {
        return Err("只有 TXT 文档支持保存。".to_owned());
    }

    let knowledge_base = snapshot
        .knowledge_bases
        .iter()
        .find(|item| item.id == snapshot.documents[document_index].knowledge_base_id)
        .ok_or_else(|| "找不到文档所属知识库".to_owned())?;
    let knowledge_base_id = knowledge_base.id.clone();
    let document_relative_path = snapshot.documents[document_index].path.clone();
    let target_path = storage::resolve_existing_file_inside_root(
        PathBuf::from(&knowledge_base.path).as_path(),
        &document_relative_path,
    )?;

    let read_path = target_path.clone();
    let current_content = run_blocking("读取待保存 TXT 文件", move || {
        fs::read_to_string(&read_path).map_err(|error| format!("无法读取待保存 TXT 文件：{error}"))
    })
    .await?;
    let current_hash = storage::hash_content(&current_content);

    // expectedHash 来自用户开始编辑时的文件版本；不一致说明外部编辑器已改动，必须先重扫。
    if current_hash != payload.expected_hash {
        logging::write_app_event_best_effort(
            &app,
            AppEventBuilder::new(
                AppLogLevel::Warn,
                AppLogCategory::Editor,
                "save_document_content",
                "blocked",
                "目标 TXT 文件已被外部修改，已阻止保存。",
            )
            .duration(started_at.elapsed())
            .knowledge_base_id(knowledge_base_id.clone())
            .entity("document", payload.document_id.clone())
            .relative_path(document_relative_path.clone()),
        );

        return Err("目标文件已被外部修改，已阻止保存。请重新扫描后再编辑。".to_owned());
    }

    capture_document_history_before_write(
        &app,
        storage::DocumentHistoryCapture {
            target_kind: "document".to_owned(),
            knowledge_base_id: knowledge_base_id.clone(),
            target_id: payload.document_id.clone(),
            relative_path: document_relative_path.clone(),
            title: snapshot.documents[document_index].title.clone(),
            file_type: "txt".to_owned(),
            content: current_content,
            source: "manual-save".to_owned(),
            session_id: None,
            change_id: None,
            operation_id: None,
        },
        AppLogCategory::Editor,
        "save_document_content",
        started_at,
    )
    .await?;

    let write_path = target_path.clone();
    let write_content = payload.content.clone();

    run_blocking("保存 TXT 文件", move || {
        storage::atomic_write_text_document(&write_path, &write_content)
    })
    .await?;

    let updated_at = read_file_updated_at_or_now(
        &app,
        "save_document_content",
        &knowledge_base_id,
        "document",
        &payload.document_id,
        &document_relative_path,
        &target_path,
    );
    let next_hash = storage::hash_content(&payload.content);
    snapshot.documents[document_index].content = Some(payload.content);
    snapshot.documents[document_index].content_hash = next_hash;
    snapshot.documents[document_index].updated_at = updated_at;
    snapshot.active_note_id.clear();
    snapshot.active_document_id = payload.document_id;
    normalize_active_entities(&mut snapshot, None);

    logging::write_app_event_best_effort(
        &app,
        AppEventBuilder::new(
            AppLogLevel::Info,
            AppLogCategory::Editor,
            "save_document_content",
            "completed",
            "已保存 TXT 文档。",
        )
        .duration(started_at.elapsed())
        .knowledge_base_id(knowledge_base_id)
        .entity("document", snapshot.active_document_id.clone())
        .relative_path(document_relative_path),
    );

    Ok(snapshot)
}

/** 加载 docx/pdf/图片文档预览，命令层负责定位知识库并把路径授权给 asset protocol。 */
#[tauri::command]
pub async fn load_document_preview(
    app: AppHandle,
    payload: LoadDocumentPreviewPayload,
) -> Result<DocumentPreview, String> {
    let snapshot = payload.snapshot;
    let document = snapshot
        .documents
        .iter()
        .find(|document| document.id == payload.document_id)
        .cloned()
        .ok_or_else(|| "找不到要预览的文档".to_owned())?;
    let knowledge_base = snapshot
        .knowledge_bases
        .iter()
        .find(|item| item.id == document.knowledge_base_id)
        .cloned()
        .ok_or_else(|| "找不到文档所属知识库".to_owned())?;
    let root_path = PathBuf::from(&knowledge_base.path);

    allow_asset_protocol_directory(&app, &root_path)?;

    run_blocking("加载文档预览", move || {
        storage::load_document_preview(&root_path, &document)
    })
    .await
}
